### ISSUE TITLE
Please provide a method to tell compiler a variable has been assigned

### ISSUE THREAD

**Issue by vipcxj** [Microsoft/TypeScript#30985]
``` javascript
let a: boolean;
let b: boolean;
let c: number;
if (a || b) {
   c = 1;
}
if (b) {
   method(c) // TS2454 here
}
```
`c` has been assigned with 1. but compiler doesn't know it. we need a method to tell compiler that, just like using `!` to tell compile a variable is not optional.

**Comment by vipcxj** [comment]
I found `!` is that method, so close it.

**Comment by vipcxj** [comment]
I want to reopen it. Or open a new issue?
`!` not work.
A new error is reported when using `!`
`This assertion is unnecessary since it does not change the type of the expression.`

**Comment by RyanCavanaugh** [comment]
@vipcxj that error is from tslint; probably something isn't right with their rule
